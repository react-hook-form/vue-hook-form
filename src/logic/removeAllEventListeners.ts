import isHTMLElement from '../utils/isHTMLElement';
import { EVENTS } from '../constants';
import { Ref } from '../types/form';

export default (
  ref: Ref,
  validateWithStateUpdate: EventListenerOrEventListenerObject,
): void => {
  if (isHTMLElement(ref) && ref.removeEventListener) {
    ref.removeEventListener(EVENTS.INPUT, validateWithStateUpdate);
    ref.removeEventListener(EVENTS.CHANGE, validateWithStateUpdate);
    ref.removeEventListener(EVENTS.BLUR, validateWithStateUpdate);
  }
};
