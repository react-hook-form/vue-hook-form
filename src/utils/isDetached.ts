import { Ref } from '../types/form';

export default function isDetached(element: Ref): boolean {
  if (!element) {
    return true;
  }

  if (
    !(element instanceof HTMLElement) ||
    element.nodeType === Node.DOCUMENT_NODE
  ) {
    return false;
  }

  return isDetached(element.parentNode as Ref);
}
