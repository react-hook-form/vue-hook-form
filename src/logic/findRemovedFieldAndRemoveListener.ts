import * as React from 'react';
import removeAllEventListeners from './removeAllEventListeners';
import getFieldValue from './getFieldValue';
import isRadioInput from '../utils/isRadioInput';
import isCheckBoxInput from '../utils/isCheckBoxInput';
import isDetached from '../utils/isDetached';
import isArray from '../utils/isArray';
import unset from '../utils/unset';
import unique from '../utils/unique';
import isUndefined from '../utils/isUndefined';
import { Field, FieldRefs, FieldValues, Ref } from '../types/form';

const isSameRef = (fieldValue: Field, ref: Ref) =>
  fieldValue && fieldValue.ref === ref;

export default function findRemovedFieldAndRemoveListener<
  TFieldValues extends FieldValues
>(
  fieldsRef: React.MutableRefObject<FieldRefs<TFieldValues>>,
  handleChange: ({ type, target }: Event) => Promise<void | boolean>,
  field: Field,
  unmountFieldsStateRef: React.MutableRefObject<Record<string, any>>,
  shouldUnregister?: boolean,
  forceDelete?: boolean,
): void {
  const {
    ref,
    ref: { name, type },
    mutationWatcher,
  } = field;
  const fieldRef = fieldsRef.current[name] as Field;

  if (!shouldUnregister) {
    const value = getFieldValue(fieldsRef, name, unmountFieldsStateRef);

    if (!isUndefined(value)) {
      unmountFieldsStateRef.current[name] = value;
    }
  }

  if (!type) {
    delete fieldsRef.current[name];
    return;
  }

  if ((isRadioInput(ref) || isCheckBoxInput(ref)) && fieldRef) {
    const { options } = fieldRef;

    if (isArray(options) && options.length) {
      unique(options).forEach((option, index): void => {
        const { ref, mutationWatcher } = option;
        if ((ref && isDetached(ref) && isSameRef(option, ref)) || forceDelete) {
          removeAllEventListeners(ref, handleChange);

          if (mutationWatcher) {
            mutationWatcher.disconnect();
          }

          unset(options, `[${index}]`);
        }
      });

      if (options && !unique(options).length) {
        delete fieldsRef.current[name];
      }
    } else {
      delete fieldsRef.current[name];
    }
  } else if ((isDetached(ref) && isSameRef(fieldRef, ref)) || forceDelete) {
    removeAllEventListeners(ref, handleChange);

    if (mutationWatcher) {
      mutationWatcher.disconnect();
    }

    delete fieldsRef.current[name];
  }
}
