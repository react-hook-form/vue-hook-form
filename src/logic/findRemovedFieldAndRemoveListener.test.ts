// @ts-nocheck
import findRemovedFieldAndRemoveListener from './findRemovedFieldAndRemoveListener';
import isDetached from '../utils/isDetached';
import removeAllEventListeners from './removeAllEventListeners';

jest.mock('./removeAllEventListeners');
jest.mock('../utils/isDetached');

describe('findMissDomAndClean', () => {
  beforeEach(() => {
    jest.resetAllMocks();

    (isDetached as any).mockImplementation(() => {
      return true;
    });
  });

  describe('radio', () => {
    it('should return default fields value if nothing matches', () => {
      document.body.contains = jest.fn(() => true);
      const fields = {
        current: {
          test: 'test',
        },
      };
      expect(
        findRemovedFieldAndRemoveListener(
          fields as any,
          () => ({} as any),
          {
            ref: { name: 'bill', type: 'radio' },
          },
          {},
          true,
        ),
      ).toEqual(undefined);
    });

    it('should remove options completely if option found and no option left', () => {
      document.body.contains = jest.fn(() => false);

      const ref = document.createElement('input');
      ref.setAttribute('name', 'test');
      ref.setAttribute('type', 'radio');

      const disconnect = jest.fn();
      const fields = {
        current: {
          test: {
            name: 'test',
            ref,
            options: [
              {
                ref,
                mutationWatcher: {
                  disconnect,
                },
              },
            ],
          },
        },
      };

      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        {
          ref,
        },
        {},
        true,
      );

      expect(fields).toEqual({
        current: {},
      });
    });

    it('should remove none radio field when found', () => {
      const ref = document.createElement('input');
      ref.setAttribute('name', 'test');
      ref.setAttribute('type', 'radio');
      document.body.contains = jest.fn(() => false);
      const disconnect = jest.fn();
      const fields = {
        current: {
          test: {
            name: 'test',
            ref: {},
            mutationWatcher: {
              disconnect,
            },
          },
          test1: {
            name: 'test',
            ref: {},
          },
        },
      };

      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        {
          ref,
          mutationWatcher: {
            disconnect,
          },
        },
        {},
      );

      expect(fields).toMatchSnapshot();
    });

    it('should work for radio type input', () => {
      const ref = document.createElement('input');
      ref.setAttribute('name', 'test');
      ref.setAttribute('type', 'radio');
      document.body.contains = jest.fn(() => false);
      const disconnect = jest.fn();
      const fields = {
        current: {
          test: {
            name: 'test',
            ref: {},
            mutationWatcher: {
              disconnect,
            },
          },
          test1: {
            name: 'test',
            ref: {
              type: 'radio',
            },
          },
        },
      };

      expect(
        findRemovedFieldAndRemoveListener(
          fields,
          () => ({} as any),
          {
            ref: { name: 'test', type: 'radio' },
            options: [{ ref }],
            mutationWatcher: {
              disconnect,
            },
          },
          {},
        ),
      ).toMatchSnapshot();
    });

    it('should work for checkbox type input', () => {
      const ref = document.createElement('input');
      ref.setAttribute('name', 'test');
      ref.setAttribute('type', 'checkbox');
      document.body.contains = jest.fn(() => false);
      const disconnect = jest.fn();
      const fields = {
        current: {
          test: {
            name: 'test',
            ref: {},
            mutationWatcher: {
              disconnect,
            },
          },
          test1: {
            name: 'test',
            ref: {
              type: 'checkbox',
            },
          },
        },
      };

      expect(
        findRemovedFieldAndRemoveListener(
          fields,
          () => ({} as any),
          {
            ref: { name: 'test', type: 'checkbox' },
            options: [{ ref }],
            mutationWatcher: {
              disconnect,
            },
          },
          {},
        ),
      ).toMatchSnapshot();
    });

    it('should not delete field when option have value', () => {
      (isDetached as any).mockImplementation(() => {
        return false;
      });

      const fields = {
        current: {
          test: {
            name: 'test',
            type: 'radio',
            ref: {},
            options: [{ ref: { name: 'test', type: 'radio' } }],
          },
        },
      };

      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        {
          ref: { name: 'test', type: 'radio' },
        },
        {},
      );

      expect(fields).toEqual({
        current: {
          test: {
            name: 'test',
            type: 'radio',
            ref: {},
            options: [{ ref: { name: 'test', type: 'radio' } }],
          },
        },
      });
    });

    it('should not remove event listener when type is not Element', () => {
      (isDetached as any).mockImplementation(() => {
        return false;
      });
      document.body.contains = jest.fn(() => false);

      const disconnect = jest.fn();
      const fields = {
        current: {
          test: {
            name: 'test',
            type: 'radio',
            ref: {},
            options: [
              {
                ref: 'test',
                mutationWatcher: {
                  disconnect,
                },
              },
            ],
          },
        },
      };

      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        {
          ref: { name: 'test', type: 'text' },
          options: [
            {
              mutationWatcher: {
                disconnect,
              },
              ref: {},
            },
          ],
        },
        {},
      );

      expect(fields).toMatchSnapshot();

      expect(
        findRemovedFieldAndRemoveListener(
          fields,
          () => ({} as any),
          {
            ref: { name: 'test', type: 'text' },
          },
          {},
        ),
      ).toMatchSnapshot();
    });

    it('should remove options when force delete is set to true', () => {
      (isDetached as any).mockImplementation(() => {
        return false;
      });

      document.body.contains = jest.fn(() => false);

      const ref = document.createElement('input');
      ref.setAttribute('name', 'test');
      ref.setAttribute('type', 'radio');

      const disconnect = jest.fn();
      const fields = {
        current: {
          test: {
            name: 'test',
            ref: {},
            options: [],
          },
        },
      };
      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        {
          ref: { name: 'test', type: 'radio' },
          options: [
            {
              mutationWatcher: {
                disconnect,
              },
              ref,
            },
          ],
        },
        {},
        false,
        true,
      );

      expect(fields).toEqual({
        current: {},
      });
    });
  });

  describe('text', () => {
    it('should delete field if type is text', () => {
      const mockWatcher = jest.fn();
      const state = { current: {} };
      const fields = {
        current: {
          test: {
            name: 'test',
            ref: {
              name: 'test',
              type: 'text',
              value: 'test',
            },
            mutationWatcher: {
              disconnect: mockWatcher,
            },
          },
        },
      };

      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        fields.current.test,
        state,
      );

      expect(state).toEqual({
        current: { test: 'test' },
      });
      expect(mockWatcher).toBeCalled();
      expect(fields).toEqual({
        current: {},
      });
    });

    it('should delete field if forceDelete is true', () => {
      (isDetached as any).mockReturnValue(false);
      const state = { current: {} };
      const fields = {
        current: {
          test: {
            name: 'test',
            ref: {
              name: 'test',
              type: 'text',
              value: 'test',
            },
          },
        },
      };

      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        fields.current.test,
        state,
        false,
        true,
      );

      expect(state).toEqual({
        current: { test: 'test' },
      });
      expect(removeAllEventListeners).toBeCalled();
      expect(fields).toEqual({
        current: {},
      });
    });

    it('should store state when component is getting unmount', () => {
      const state = { current: {} };
      const fields = {
        current: {
          test: {
            name: 'test',
            ref: {
              value: 'test',
            },
          },
        },
      };

      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        {
          ref: { name: 'test', type: 'text' },
        },
        state,
        false,
      );

      expect(state).toEqual({
        current: { test: 'test' },
      });
    });

    it('should not store state when component is getting unmount and value is return undefined', () => {
      const state = { current: {} };
      const fields = {
        current: {
          test: {
            name: 'test',
            ref: {},
          },
        },
      };

      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        {
          ref: { name: 'test', type: 'text' },
        },
        state,
        false,
      );

      expect(state).toEqual({
        current: {},
      });
    });
  });

  it('should not call mutation watcher when not available', () => {
    jest.spyOn(document.body, 'contains').mockReturnValue(false);

    const ref = document.createElement('input');
    ref.setAttribute('name', 'test');
    ref.setAttribute('type', 'radio');

    const fields = {
      current: {
        test: {
          name: 'test',
          ref,
          options: [
            {
              ref,
            },
          ],
        },
      },
    };

    expect(() => {
      findRemovedFieldAndRemoveListener(
        fields,
        () => ({} as any),
        {
          ref,
        },
        {},
        true,
      );
    }).not.toThrow();

    document.body.contains.mockRestore();
  });
});
