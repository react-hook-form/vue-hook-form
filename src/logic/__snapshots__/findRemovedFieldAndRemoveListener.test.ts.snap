// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findMissDomAndClean radio should not remove event listener when type is not Element 1`] = `
Object {
  "current": Object {
    "test": Object {
      "name": "test",
      "options": Array [
        Object {
          "mutationWatcher": Object {
            "disconnect": [MockFunction],
          },
          "ref": "test",
        },
      ],
      "ref": Object {},
      "type": "radio",
    },
  },
}
`;

exports[`findMissDomAndClean radio should not remove event listener when type is not Element 2`] = `undefined`;

exports[`findMissDomAndClean radio should remove none radio field when found 1`] = `
Object {
  "current": Object {
    "test1": Object {
      "name": "test",
      "ref": Object {},
    },
  },
}
`;

exports[`findMissDomAndClean radio should work for checkbox type input 1`] = `undefined`;

exports[`findMissDomAndClean radio should work for radio type input 1`] = `undefined`;
